### PR TITLE
SortingStep: deduce way to sort based on input stream sort description

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Deduce way to sort based on input stream sort description. Skip sorting if input stream is already sorted

### CODE REVIEW

**Comment by devcrafter** [comment]
@Mergifyio update

File: src/Processors/QueryPlan/Optimizations/reuseStorageOrderingForWindowFunctions.cpp [hunk 1]
```diff
@@ -102,10 +102,7 @@ size_t tryReuseStorageOrderingForWindowFunctions(QueryPlan::Node * parent_node,
             limit);
 
     if (order_info)
-    {
         read_from_merge_tree->setQueryInfoInputOrderInfo(order_info);
-        sorting->convertToFinishSorting(order_info->order_key_prefix_descr);
```
Comment (devcrafter) [inline src/Processors/QueryPlan/Optimizations/reuseStorageOrderingForWindowFunctions.cpp:107]: It shouldn't be necessary any more, - we should get sort description from input stream on sorting step. @novikd please could you have a look?

File: tests/performance/sorting_step_sort_descr_adoption.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<test>
```
Comment (devcrafter) [inline tests/performance/sorting_step_sort_descr_adoption.xml:1]: Probably rename file according to setting name

**Comment by devcrafter** [comment]
@nickitat @KochetovNicolai would you mind to review?

**Comment by nickitat** [comment]
some of the possible bad cases from the chat with Vladimir
https://pastila.nl/?009cde16/669061285f5478a836a4837d2b41ca00
added just to not be lost from the context

**Comment by nickitat** [comment]
`distinct_in_order` perf test seem to be flaky

File: src/Processors/QueryPlan/SortingStep.cpp [hunk 3]
```diff
@@ -111,129 +114,186 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
...
-                    pipeline.getHeader(),
-                    pipeline.getNumStreams(),
-                    result_description,
-                    max_block_size,
-                    SortingQueueStrategy::Batch,
-                    limit);
-
-            pipeline.addTransform(std::move(transform));
+            LOG_DEBUG(getLogger(), "Almost FullSort");
+            fullSort(pipeline, result_description, limit, true);
```
Comment (nickitat) [inline src/Processors/QueryPlan/SortingStep.cpp:290]: imo, this way would be more readable:
``` cpp
const bool partial_sort_not_needed = ...;
fullSort(pipeline, result_description, limit, partial_sort_not_needed);
```
without if

File: src/Processors/QueryPlan/SortingStep.cpp [hunk 4]
```diff
@@ -111,129 +114,186 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
...
+                remerge_lowered_memory_bytes_ratio,
+                max_bytes_before_external_sort,
+                tmp_volume,
+                min_free_disk_space);
         });
+}
+
+void SortingStep::fullSort(QueryPipelineBuilder & pipeline, const SortDescription & result_sort_desc, const UInt64 limit_, const bool skip_partial_sort)
+{
+    if (!skip_partial_sort)
```
Comment (nickitat) [inline src/Processors/QueryPlan/SortingStep.cpp:199]: I'm not sure that it will not introduce regression in some cases. `PartialSortingTransform` can throw out significant portion of rows due to the `limit`. pls, check it carefully (also taking [this](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Transforms/PartialSortingTransform.h#L44) into account), let's add a perf test for such case.

File: src/Interpreters/ActionsDAG.cpp [hunk 5]
```diff
@@ -1923,4 +1924,34 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
     return actions;
 }
 
+bool ActionsDAG::isSortingPreserved() const
```
Comment (KochetovNicolai) [inline src/Interpreters/ActionsDAG.cpp:1927]: This property is a little bit strange.
It may be reasonable to make it dependent on some inputs and return outputs.

It is not very often that function is globally monotonic.
Even if all the functions are globally monotonic, we can remove a sorting key column from result.

File: src/Processors/QueryPlan/ExpressionStep.cpp [hunk 6]
```diff
@@ -5,35 +5,40 @@
...
         },
         {
             .preserves_number_of_rows = !actions->hasArrayJoin(),
-        }
-    };
+        }};
 }
 
-ExpressionStep::ExpressionStep(const DataStream & input_stream_, ActionsDAGPtr actions_dag_)
+ExpressionStep::ExpressionStep(const DataStream & input_stream_, const ActionsDAGPtr & actions_dag_)
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/ExpressionStep.cpp:33]: Why?

File: src/Processors/QueryPlan/ExpressionStep.cpp [hunk 7]
```diff
@@ -5,35 +5,40 @@
...
 #include <Interpreters/JoinSwitcher.h>
-
 #include <Common/JSONBuilder.h>
 
 namespace DB
 {
 
+static Poco::Logger * getLogger()
+{
+    static Poco::Logger & logger = Poco::Logger::get("ExpressionStep");
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/ExpressionStep.cpp:15]: I think it's not a good idea to store logger statically.

File: src/Processors/QueryPlan/SortingStep.h [hunk 8]
```diff
@@ -49,25 +50,35 @@ class SortingStep : public ITransformingStep
...
+    void mergeSorting(QueryPipelineBuilder & pipeline, const SortDescription & result_sort_desc, UInt64 limit_);
+    void finishSorting(
+        QueryPipelineBuilder & pipeline, const SortDescription & input_sort_desc, const SortDescription & result_sort_desc, UInt64 limit_);
+    void fullSort(
+        QueryPipelineBuilder & pipeline,
+        const SortDescription & result_sort_desc,
+        UInt64 limit_,
+        bool skip_partial_sort = false);
+
     enum class Type
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/SortingStep.h:70]: Agree. 
I think that Full was better then Auto
Also I think we would better remove type at all

File: src/Interpreters/InterpreterExplainQuery.cpp [hunk 9]
```diff
@@ -174,7 +174,8 @@ struct QueryPlanSettings
             {"actions", query_plan_options.actions},
             {"indexes", query_plan_options.indexes},
             {"optimize", optimize},
-            {"json", json}
+            {"json", json},
+            {"sortmode", query_plan_options.sort_mode},
```
Comment (KochetovNicolai) [inline src/Interpreters/InterpreterExplainQuery.cpp:178]: I think it may be more reasonable to add a common setting which shows all the known data stream properties.
Right now it can be sort description + sort mode, distinct columns

File: src/Processors/QueryPlan/SortingStep.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Processors/QueryPlan/SortingStep.h:70]: When `SortingStep` constructed with `MergingSorted` or `FinishSorting`, - we consider that a caller has more information what sorting to apply, so we'll not try to infer it. And `Auto` is a mode when we infer the way to sort based on sorting properties of input stream. I'll make a comment

File: src/Core/SortDescription.cpp [hunk 10]
```diff
@@ -42,6 +42,19 @@ void SortColumnDescription::explain(JSONBuilder::JSONMap & map) const
     map.add("With Fill", with_fill);
 }
 
+bool SortDescription::hasPrefix(const SortDescription & prefix) const
+{
+    if (prefix.empty() || prefix.size() > size())
+        return false;
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Core/SortDescription.cpp]: ok

File: src/Core/SortDescription.cpp [hunk 11]
```diff
@@ -42,6 +42,19 @@ void SortColumnDescription::explain(JSONBuilder::JSONMap & map) const
     map.add("With Fill", with_fill);
 }
 
+bool SortDescription::hasPrefix(const SortDescription & prefix) const
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Core/SortDescription.cpp:45]: Don't see how it fits. Prefix and description itself don't always have the same number of elements

File: src/Interpreters/ActionsDAG.cpp [hunk 12]
```diff
@@ -1926,4 +1927,21 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
...
+    if (sort_description.empty())
+        return true;
+
+    if (hasArrayJoin())
+        return false;
+
+    const Block & output_header = updateHeader(header);
+    for (const auto & desc : sort_description)
+    {
+        if (!output_header.findByName(desc.column_name))
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Interpreters/ActionsDAG.cpp:1941]: Yes, it's not covered yet

**Comment by qoega** [comment]
02377_optimize_sorting_for_input_stream_explain - need to update reference

File: src/Processors/QueryPlan/SortingStep.cpp [hunk 13]
```diff
@@ -57,14 +60,14 @@ SortingStep::SortingStep(
 
 SortingStep::SortingStep(
     const DataStream & input_stream_,
-    SortDescription prefix_description_,
-    SortDescription result_description_,
+    const SortDescription & prefix_description_,
+    const SortDescription & result_description_,
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/SortingStep.cpp:64]: Why this change is needed? Passing by value + std::move will avoid potential copy.

Comment (KochetovNicolai) [reply src/Processors/QueryPlan/SortingStep.cpp:64]: For example, here, if we add std::move to `order_descr`
https://github.com/ClickHouse/ClickHouse/blob/fe6533f41db18e7bb75a29f660c0fb2f68ee98d7/src/Interpreters/InterpreterSelectQuery.cpp#L1438

File: src/Processors/QueryPlan/SortingStep.cpp [hunk 14]
```diff
@@ -75,12 +78,12 @@ SortingStep::SortingStep(
 
 SortingStep::SortingStep(
     const DataStream & input_stream,
-    SortDescription sort_description_,
+    const SortDescription & sort_description_,
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/SortingStep.cpp:81]: .

File: src/Processors/QueryPlan/SortingStep.cpp [hunk 15]
```diff
@@ -111,129 +114,174 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
...
-            auto transform = std::make_shared<MergingSortedTransform>(
-                    pipeline.getHeader(),
-                    pipeline.getNumStreams(),
-                    result_description,
-                    max_block_size,
-                    SortingQueueStrategy::Batch,
-                    limit);
-
-            pipeline.addTransform(std::move(transform));
+            type = Type::MergingSorted;
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/SortingStep.cpp:270]: Right now we never call transformPipeline more than once, but there may be good to allow it for some optimizations.
Let's try not to change internal state here.
(Maybe I should have done transformPipeline const)

File: src/Processors/QueryPlan/SortingStep.cpp (hunk 13 above)
Comment (devcrafter) [reply src/Processors/QueryPlan/SortingStep.cpp]: Agree. It was oversimplification

File: src/Processors/QueryPlan/QueryPlan.cpp [hunk 16]
```diff
@@ -325,8 +325,22 @@ static void explainStep(
...
-
         settings.out.write('\n');
+
+    }
+
+    if (options.sort_mode)
+    {
+        if (step.hasOutputStream())
+        {
+            settings.out << prefix << "Sort Mode: " << step.getOutputStream().sort_mode;
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Processors/QueryPlan/QueryPlan.cpp:336]: imo, the sorting properties are logically one thing (scope + how sorted). Internal separation into different variables is implementation detail. So, I'd keep it at one line. Regarding the formatting of this line - we can change it.
Also, one line allows checking the sorting in `explain` test much easier, - there is no need to check multiple lines.

@KochetovNicolai @vdimir opinions?

File: src/Interpreters/ActionsDAG.cpp [hunk 17]
```diff
@@ -1926,4 +1927,49 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
...
+            if (!output_header.findByName(it->second))
+                return false;
+        }
+
+        /// check that found colunm is not an alias
+        for (const Node * node : index)
+        {
+            if (node->type == ActionType::ALIAS && node->result_name == desc.column_name)
+                return false;
+        }
```
Comment (KochetovNicolai) [inline src/Interpreters/ActionsDAG.cpp:1970]: I will try to create a counter-example for this

**Review by KochetovNicolai — COMMENTED** [review]
Generally GLTM, but I will need to check `ActionsDAG::isSortingPreserved`

**Comment by KochetovNicolai** [comment]
Found.
```
create table tab (x UInt64) engine = MergeTree order by x + 1;
insert into tab select number from numbers(4);
select x, y from (select x + 1 as y, sipHash64(x) as x from (select x from tab order by x + 1)) order by x + 1;

SELECT
    x,
    y
FROM
(
    SELECT
        x + 1 AS y,
        sipHash64(x) AS x
    FROM
    (
        SELECT x
        FROM tab
        ORDER BY x + 1 ASC
    )
)
ORDER BY x + 1 ASC

Query id: a032735e-1364-4193-8156-6511b478dd67

┌────────────────────x─┬────────────────────y─┐
│ 16738165381834614119 │ 16738165381834614120 │
│  9224715256000962398 │  9224715256000962399 │
│ 13686418376000424449 │ 13686418376000424450 │
│ 14247234200463187066 │ 14247234200463187067 │
└──────────────────────┴──────────────────────┘
```

```
set optimize_sorting_for_input_stream=0;
select x, y from (select x + 1 as y, sipHash64(x) as x from (select x from tab order by x + 1)) order by x + 1;

SELECT
    x,
    y
FROM
(
    SELECT
        x + 1 AS y,
        sipHash64(x) AS x
    FROM
    (
        SELECT x
        FROM tab
        ORDER BY x + 1 ASC
    )
)
ORDER BY x + 1 ASC

Query id: d4611d8c-890c-4ce0-82d5-396ed61b5592

┌────────────────────x─┬────────────────────y─┐
│  9224715256000962398 │  9224715256000962399 │
│ 13686418376000424449 │ 13686418376000424450 │
│ 14247234200463187066 │ 14247234200463187067 │
│ 16738165381834614119 │ 16738165381834614120 │
└──────────────────────┴──────────────────────┘
```

**Comment by devcrafter** [comment]
@KochetovNicolai Please check the fix. It addresses the case you provided

**Comment by devcrafter** [comment]
@KochetovNicolai @nickitat Please could you review it once more? It should be ready. I've added comments regarding possible follow-ups

File: tests/queries/0_stateless/02377_optimize_sorting_for_input_stream_explain.reference [hunk 18]
```diff
@@ -37,7 +37,7 @@ Sort Mode: Stream: x ASC, y ASC
 Sort Mode: Chunk: a ASC, b ASC
 Sort Mode: Chunk: a ASC, b ASC
 -- PLAN: actions chain breaks sorting order: input(column a)->sipHash64(column a)->alias(sipHash64(column a), a)->plus(alias a, 1)
-Sort Mode: Stream: plus(a, 1) ASC
+Sort Mode: None
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply tests/queries/0_stateless/02377_optimize_sorting_for_input_stream_explain.reference:64]: (1) I think it's ok. Would be perfect to propagate this information, but it's not a error if we can't.
(2) Hard to say why it is happening. Probably some logic which removes columns uses only names. For a query plan (and ActionsDAG) it is allowed to have duplicate columns (duplicates are resolved in-order).

File: src/Interpreters/ActionsDAG.cpp [hunk 19]
```diff
@@ -1941,4 +1942,102 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
...
+        if (output_header.findByName(desc.column_name))
+        {
+            /// find the corresponding node in output
+            const auto * output_node = tryFindInOutputs(desc.column_name);
+            if (!output_node)
+            {
+                /// sorted column name in header but NOT in expression output -> no expression is applied to it -> sorting preserved
+                continue;
+            }
+        }
```
Comment (KochetovNicolai) [inline src/Interpreters/ActionsDAG.cpp:2018]: This code is ok.
I think it is the same as to check if there is an input with this name.

**Review by KochetovNicolai — APPROVED** [review]

File: tests/queries/0_stateless/02377_optimize_sorting_for_input_stream_explain.reference (hunk 18 above)
[the remark this answers is not in the document]
Comment (devcrafter) [reply tests/queries/0_stateless/02377_optimize_sorting_for_input_stream_explain.reference:64]: I'll try to address it later as follow-up tasks

File: src/Processors/QueryPlan/QueryPlan.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/QueryPlan.cpp:336]: Generally agree with Nikita.
Also I would not print anything if sort mode is None

[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Processors/QueryPlan/QueryPlan.cpp:336]: Discussed with @KochetovNicolai 
Agreed to the following formating (one line):
```
Sorting (<scope>): <sort description>
```

[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Processors/QueryPlan/QueryPlan.cpp:336]: Also, regarding sorting scope (mode), we agreed that current scopes `Port` and `Stream` are confusing.
The proposal is to rename them (as follow-up):
`Port` -> `Stream`
`Stream` -> `Full` (or `Global`)

@nickitat opinion?

File: src/Interpreters/ActionsDAG.cpp [hunk 20]
```diff
@@ -1941,4 +1942,102 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
...
+                if (!func->hasInformationAboutMonotonicity())
+                    return false;
+
+                const auto & types = func->getArgumentTypes();
+                if (types.empty())
+                    return false;
+
+                /// TODO: we support monotonicity check only for functions with one parameter but ...
+                ///       if one parameter is variable and other are constant then we can try to check monotonicity as well
+                const auto monotonicity = func->getMonotonicityForRange(*types.front(), field, field);
```
Comment (nickitat) [inline src/Interpreters/ActionsDAG.cpp:1965]: in all three cases: here, at l.1974 and l.1991 imo should be provided some comments, why only `front()` is considered.

**Comment by devcrafter** [comment]
> "global" is a good name. "stream", I am afraid, will lead to confusion between DataStream and streams on pipeline level. imo "port" is ok. and "stream" also ok for me.

It looks to me as opposite. Regarding parallel processing, we're saying that we have several data streams processed in parallel - if each stream is sorted, then sorting scope (we call it sort mode) is _stream_. Port is pipeline level implementation detail.

File: tests/queries/0_stateless/02377_optimize_sorting_for_input_stream_explain.reference (hunk 18 above)
[the remark this answers is not in the document]
Comment (nickitat) [reply tests/queries/0_stateless/02377_optimize_sorting_for_input_stream_explain.reference]: btw, we should be careful with functions like `plus`: `18446744073709551614 < 18446744073709551615`, but if we apply `plus(*, 1)` it became false.

File: src/Core/Settings.h [hunk 21]
```diff
@@ -626,6 +626,7 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(Bool, compatibility_ignore_auto_increment_in_create_table, false, "Ignore AUTO_INCREMENT keyword in column declaration if true, otherwise return error. It simplifies migration from MySQL", 0) \
     M(Bool, multiple_joins_try_to_keep_original_names, false, "Do not add aliases to top level expression list on multiple joins rewrite", 0) \
     M(Bool, optimize_distinct_in_order, true, "Enable DISTINCT optimization if some columns in DISTINCT form a prefix of sorting. For example, prefix of sorting key in merge tree or ORDER BY statement", 0) \
+    M(Bool, optimize_sorting_for_input_stream, true, "Optimize sorting by sorting properties of input stream", 0) \
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Core/Settings.h:629]: Then probably `optimize_sorting_by_input_stream_sorting_properties` ?

**Review by nickitat — APPROVED** [review]

**Comment by devcrafter** [comment]
Stress test failure is known issue #39857, see https://s3.amazonaws.com/clickhouse-test-reports/38719/fdd699e8a377c25bf8ff9e1d2c14f760cc79a4ee/stress_test__undefined_/fatal_messages.txt

**Comment by devcrafter** [comment]
> > "global" is a good name. "stream", I am afraid, will lead to confusion between DataStream and streams on pipeline level. imo "port" is ok. and "stream" also ok for me.
> 
> It looks to me as opposite. Regarding parallel processing, we're saying that we have several data streams processed in parallel - if each stream is sorted, then sorting scope (we call it sort mode) is _stream_. Port is pipeline level implementation detail.

See proposal for renaming #40305
